protocols/perf/src/server/handler.rs

```diff
@@ -0,0 +1,158 @@
+// Copyright 2023 Protocol Labs.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use std::{
+    task::{Context, Poll},
+    time::{Duration, Instant},
+};
+
+use futures::{future::BoxFuture, stream::FuturesUnordered, FutureExt, StreamExt};
+use libp2p_core::upgrade::{DeniedUpgrade, ReadyUpgrade};
+use libp2p_swarm::{
+    handler::{
+        ConnectionEvent, DialUpgradeError, FullyNegotiatedInbound, FullyNegotiatedOutbound,
+        ListenUpgradeError,
+    },
+    ConnectionHandler, ConnectionHandlerEvent, ConnectionHandlerUpgrErr, KeepAlive,
+    SubstreamProtocol,
+};
+use void::Void;
+
+use super::RunStats;
+
+#[derive(Debug)]
+pub enum Event {
+    Finished { stats: RunStats },
+}
+
+pub struct Handler {
+    inbound: FuturesUnordered<BoxFuture<'static, Result<RunStats, std::io::Error>>>,
+    keep_alive: KeepAlive,
+}
+
+impl Handler {
+    pub fn new() -> Self {
+        Self {
+            inbound: Default::default(),
+            keep_alive: KeepAlive::Yes,
+        }
+    }
+}
+
+impl Default for Handler {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ConnectionHandler for Handler {
+    type InEvent = ();
+    type OutEvent = Event;
+    type Error = std::io::Error;
```

Comment:
```suggestion
    type Error = Void;
```
 You never close the connection actively.

Reply:
d0e96d2d